rosetta-source/src/main/rosetta/event-common-func.rosetta

```diff
@@ -1661,6 +1661,173 @@ func Qualify_Shaping: <"The qualification of a shaping event from the fact that
+				empty
+				),
+			empty, empty, 
+			Create_RepriceTermsChangeInstruction(
+				tradeState -> trade -> tradableProduct -> product -> contractualProduct,
+				haircutPercentage,
+				effectiveRepriceDate
+				),
+			empty, empty, empty, empty
+			)
+		]
+		
+		
+func Create_RepriceTermsChangeInstruction: <"Creates the relevant terms change primitive instruction object for repricing a contractual product, which consists in the same terms as the original contractual product 
+											but with a new collateral price that may also result in modifing the cash or collateral quantity.">
+	inputs:
+		contractualProduct ContractualProduct (1..1) <"The original contractual product to be repriced.">
+		newHaircutPercentage number (1..1) <"The new haricut value based on cash adjustment.">
+		effectiveRepriceDate AdjustableOrRelativeDate (1..1) <"The date to close and open a new position.">
+	output:
+		termsChangeInstruction TermsChangeInstruction (1..1) <"The relevant primitive instruction for the repricing.">
+
+	set termsChangeInstruction -> product -> contractualProduct:
+		contractualProduct
+	set termsChangeInstruction -> product -> contractualProduct -> economicTerms -> collateral -> collateralProvisions -> eligibleCollateral -> criteria -> treatment ->
+		valuationTreatment -> haircutPercentage:
+		newHaircutPercentage
+	set termsChangeInstruction -> product -> contractualProduct -> economicTerms -> effectiveDate:
+		effectiveRepriceDate
+		
+func Qualify_Reprice:
+	[qualification BusinessEvent]
+	inputs:
+		businessEvent BusinessEvent (1..1)
+	output: is_event boolean (1..1)
+	alias beforeTradePrice: businessEvent -> instruction only-element -> before -> trade -> tradableProduct -> tradeLot->priceQuantity
+
+	alias openTradePrice: FilterOpenTradeStates( businessEvent -> after ) only-element -> trade -> tradableProduct -> tradeLot->priceQuantity
+
+	alias closedTradePrice: FilterClosedTradeStates( businessEvent -> after )
+	
+	alias beforeEconomicterms: businessEvent -> instruction only-element -> before -> trade -> tradableProduct -> product -> contractualProduct -> economicTerms
+	alias openEconomicTerms: FilterOpenTradeStates( businessEvent -> after ) only-element -> trade -> tradableProduct -> product -> contractualProduct -> economicTerms
+	alias closedTradeState: FilterClosedTradeStates( businessEvent -> after )
+
+	set is_event:
+		// The first 2 conditions imply that there is 1 and only 1 before and 1 open thanks to the "only-element" modifier
+		beforeTradePrice exists and
+		openTradePrice exists and
+		closedTradePrice count = 1 and
+		openTradePrice  <> beforeTradePrice and
+		businessEvent -> after -> trade -> tradableProduct -> product -> contractualProduct -> economicTerms -> payout -> assetPayout exists and
+		beforeEconomicterms exists and
+		openEconomicTerms exists and
+		closedTradeState count = 1 and
+		openEconomicTerms -> terminationDate = beforeEconomicterms -> terminationDate
```

Comment:
not sure why we need to test the termination date has not changed?

Reply:
That's correct. This is to ensure that the termination date has not changed otherwise it's more that a reprice.